System/Random/MWC/Distributions.hs

```diff
@@ -70,31 +73,185 @@ normal m s gen = do
+    (Nothing, Nothing) -> pkgError "truncatedNormalRange" "no bound returned -- use `normal` instead"
+    (Just mn, Nothing) -> pure undefined
+    (Nothing, Just mx) -> pure undefined
+    (Just mn, Just mx) ->
+      if mn > mx
+      then pkgError "truncatedNormalRange" "min bound is greater than max bound"
+      else
+        if abs mn > abs mx
+        then (* (-1)) <$> truncatedNormalRange (Just (-mx), Just (-mn)) m s gen
+        else -- If a in the right tail (a > xmax), use rejection algorithm with a truncated exponential proposal
+          if a > xmax
+          then scale <$> rtexp mn mx gen
+          else -- If a in the left tail (a < xmin), use rejection algorithm with a Gaussian proposal
+            if a < xmin
+            then scale <$> rtgaussian mn mx
+            else chopin (mn, mx) m s gen
+
+  where
+    -- Design variables
+    xmin  = -2.00443204036    :: Double             -- Left bound
+    xmax  =  3.48672170399    :: Double             -- Right bound
+    kmin  = 5                 :: Int                -- if kb-ka < kmin then use a rejection algorithm
+    invh  = 1631.73284006     :: Double             -- = 1/h, h being the minimal interval range
+    i0    = 3271              :: Int                -- = - floor(x(0)/h)
+    alpha = 1.837877066409345 :: Double             -- = log(2*pi)
+    sq2   = 7.071067811865475e-1 :: Double          -- = 1/sqrt(2)
+    sqpi  = 1.772453850905516    :: Double          -- = sqrt(pi)
+    -- int xsize=sizeof(Rtnorm::x)/sizeof(double)   -- Length of table x
+
+    scale :: Double -> Double
+    scale x = if m /= 0 || s /= 1 then (x-m) / s else x
+
+    --------------------------------------------------------------
+    -- Rejection algorithm with a truncated exponential proposal
+    rtgaussian :: Double -> Double -> m Double
+    rtgaussian mn mx =
+      acceptReject (normal m s gen) (\z -> z >= mn && z <= mx) id
+
+--------------------------------------------------------------
+-- Rejection algorithm with a truncated exponential proposal
+rtexp :: forall m . PrimMonad m => Double -> Double -> Gen (PrimState m) -> m Double
+rtexp mn mx gen = acceptRejectExp
+  where
+    twoasq :: Double
+    twoasq = 2 * (mn ^^ 2)
 
+    expab  :: Double
+    expab  = exp (-mn * (mx-mn) ) - 1
 
--- | Generate truncated normally distributed random variable
-truncatedNormalRange
-  :: PrimMonad m
-  => (Double, Double) -- ^ Range to truncate over
+    acceptRejectExp :: m Double
+    acceptRejectExp = acceptReject mkRands accept transform
+
+    mkRands :: m (Double, Double)
+    mkRands = do
+      z <- (log . (+1) . (*expab)) <$> uniform gen
+      e <- ((*(-1)) . log) <$> uniform gen
+      pure (z, e)
+
+    accept :: (Double, Double) -> Bool
+    accept (z, e) = twoasq * e > z ^^ 2
+
+    transform :: (Double, Double) -> Double
+    transform (z, _) = mn - (z / mn)
+
+
+acceptReject :: PrimMonad m => m r -> (r -> Bool) -> (r -> Double) -> m Double
+acceptReject rgen cond transform =
```

Comment:
No need for `transform` parameter. Transform could be easily added using `fmap` when needed